warehouse/utils/security_policy.py

```diff
@@ -0,0 +1,109 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import enum
+
+from pyramid.authorization import Authenticated
+from pyramid.interfaces import ISecurityPolicy
+from pyramid.security import Denied
+from zope.interface import implementer
+
+from warehouse.accounts.models import User
+
+
+class AuthenticationMethod(enum.Enum):
+    BASIC_AUTH = "basic-auth"
+    SESSION = "session"
+    MACAROON = "macaroon"
+
+
+def _principals_for_authenticated_user(user):
+    """Apply the necessary principals to the authenticated user"""
+    principals = []
+    if user.is_superuser:
+        principals.append("group:admins")
+    if user.is_moderator or user.is_superuser:
+        principals.append("group:moderators")
+    if user.is_psf_staff or user.is_superuser:
+        principals.append("group:psf_staff")
+
+    # user must have base admin access if any admin permission
+    if principals:
+        principals.append("group:with_admin_dashboard_access")
+
+    return principals
+
+
+@implementer(ISecurityPolicy)
+class MultiSecurityPolicy:
+    """
+    A wrapper for multiple Pyramid 2.0-style "security policies", which replace
+    Pyramid 1.0's separate AuthN and AuthZ APIs.
+
+    Security policies are checked in the order provided during initialization,
+    with the following semantics:
+
+    * `identity`: Selected from the first policy to return non-`None`
+    * `authenticated_userid`: Selected from the request identity, if present
+    * `forget`: Combined from all policies
+    * `remember`: Combined from all policies
+    * `permits`: Uses the AuthZ policy passed during initialization
+
+    These semantics mostly mirror those of `pyramid-multiauth`.
+    """
+
+    def __init__(self, policies, authz):
+        self._policies = policies
+        self._authz = authz
+
+    def identity(self, request):
+        for policy in self._policies:
+            if ident := policy.identity(request):
+                return ident
+
+        return None
+
+    def authenticated_userid(self, request):
+        if request.identity:
+            return str(request.identity.id)
+        return None
```

Comment:
This is fine for now but is going to be problematic once we have non-user identities, since these won't have a prefix like the principals would. We might want to address that here in advance, e.g. by asserting that the identity is a user  and otherwise returning `None` or something similar.

Reply:
Yeah, I'll make it so that only `User` identities return an ID here. That'll be consistent with our expected behavior, i.e. that `request.user` will only ever be populated with `User` models.

Reply:
[ec2c563](https://github.com/pypa/warehouse/pull/11218/commits/ec2c563d4003556435664911a23364c7bac20197)